enhancements/deprecated-bundles.md

```diff
@@ -95,32 +84,56 @@ we then deprecate version 1.3.0 using the following command
 Now, in `quay.io/my/index:v4.7` the above channel looks like:
 
 ```
-    1.4.0 -- replaces -> 1.3.0 [deprecated]
+    1.4.0 -- replaces -> 1.3.0 [deprecated] -- replaces -> 1.2.0 [deprecated] -- replaces -> 1.1.0 [deprecated] 
 ```
 
-Deprecated versions are upgradable but not installable. To achieve this, we will modify the operator-registry apis used by OLM to determine what can be installed.
+Deprecated versions are upgradable but not installable.
 
 #### Story 2
 
 As an index maintainer I want to be able to undo the above operation.
 
 ##### Implementation Details
 
-Index maintainers should revert to the index image used in the original command. If the tag is overwritten, they must keep track of the shas of the index images they modify in order to guarantee recoverability. If the image is, for some reason, not available, the index maintainer can remove the entire package using `opm index rm` and re-add all the bundles for that package.
+A flag called `--undo` is added to the deprecation command that removes the deprecated property on the bundles specified.
+
+For example, suppose following channel:
 
-Eventually, the sha of the index being modified could be stored in the database of the modified index for quicker recovery. Additionally this field would allow stepping backward with a step size equivalent to the operations / transactions provided by `opm` guaranteeing stepping back to a valid index image.
+```
+    1.4.0 -- replaces -> 1.3.0 [deprecated] -- replaces -> 1.2.0 [deprecated] -- replaces -> 1.1.0 [deprecated] 
+```
+
+we then deprecate version 1.3.0 using the following command
+
+```
+    $ opm index deprecate --bundles "quay.io/my/bundle:1.3.0" --from-index "quay.io/my/index:v4.7" --tag "quay.io/my/index:v4.7 --undo"
+```
+
+Now, in `quay.io/my/index:v4.7` the above channel looks like:
+
+```
+    1.4.0 -- replaces -> 1.3.0 -- replaces -> 1.2.0 -- replaces -> 1.1.0
+```
 
 #### Story 3
 
+Semver and semver-skipatch modes support deprecation. Adding to an index that has deprecated bundles should appropriately reject or deprecate new additions.
+
+For example, given:
+
+```
+    1.4.0 -- replaces -> 1.3.0 [deprecated] -- replaces -> 1.2.0 [deprecated] -- replaces -> 1.1.0 [deprecated] 
+```
+
+Inserting `1.2.1` into the above channel in semver mode should either produce an error or result in `1.2.1` being also marked as deprecated.
```

Comment:
can't be either or.  we're going to have to pick one.

but i'm not convinced it should be either of those.  why shouldn't it be valid to add a non-deprecated 1.2.1?  

i'm especially thinking about what may happen with freshmaker rebuilds. Which aren't using semver, but are going to be inserting new bundles into the middle of a graph that may be otherwise deprecated.

Reply:
should freshmaker rebuild deprecated bundles? I'm not sure teams will want to go through the process of releasing fixes to versions they've specifically said they don't support anymore. Maybe this is worth discussing with the teams

Reply:
they may not be deprecated in all indexes.